pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/BaseRecordExtractor.java

```diff
@@ -0,0 +1,174 @@
+    }
+    return convertedValue;
+  }
+
+  /**
+   * Returns whether the object is an instance of the data format's base type. Override this method if the extractor
+   * can handle the conversion of nested record types.
+   */
+  protected boolean isInstanceOfRecord(Object value) {
+    return false;
+  }
+
+  /**
+   * Returns whether the object is of a multi-value type. Override this method if the data format represents
+   * multi-value objects differently.
+   */
+  protected boolean isInstanceOfMultiValue(Object value) {
+    return value instanceof Collection;
+  }
+
+  /**
+   * Returns whether the object is of a map type. Override this method if the data format represents map objects
+   * differently.
+   */
+  protected boolean isInstanceOfMap(Object value) {
+    return value instanceof Map;
+  }
+
+  /**
+   * Handles the conversion of every field of the object for the particular data format. Override this method if the
+   * extractor can convert nested record types.
+   */
+  @Nullable
+  protected Object convertRecord(Object value) {
+    throw new UnsupportedOperationException("Extractor cannot convert record type structures for this data format.");
+  }
+
+  /**
+   * Handles the conversion of each element of a multi-value object. Returns {@code null} if the field value is
+   * {@code null}.
+   *
+   * This implementation converts the Collection to an Object array. Override this method if the data format
+   * requires a different conversion for its multi-value objects.
+   */
+  @Nullable
+  protected Object convertMultiValue(Object value) {
+    Collection collection = (Collection) value;
+    if (collection.isEmpty()) {
+      return null;
+    }
+
+    int numValues = collection.size();
+    Object[] array = new Object[numValues];
+    int index = 0;
+    for (Object element : collection) {
+      Object convertedValue = null;
+      if (element != null) {
+        convertedValue = convert(element);
+      }
+      if (convertedValue != null && !convertedValue.toString().equals("")) {
+        array[index++] = convertedValue;
+      }
+    }
+
+    if (index == numValues) {
+      return array;
+    } else if (index == 0) {
+      return null;
+    } else {
+      return Arrays.copyOf(array, index);
+    }
+  }
+
+  /**
+   * Handles the conversion of every value of the map. Note that map keys will be handled as a single-value type.
+   * Returns {@code null} if the field value is {@code null}. This should be overridden if the data format requires
+   * a different conversion for map values.
+   */
+  @Nullable
+  protected Object convertMap(Object value) {
+    Map map = (Map) value;
+    if (map.isEmpty()) {
+      return null;
+    }
+
+    Map<Object, Object> convertedMap = new HashMap<>();
+    for (Object key : map.keySet()) {
+      Object convertedValue = null;
+      if (key != null) {
+        convertedValue = convert(map.get(key));
+      }
+      convertedMap.put(convertSingleValue(key), convertedValue);
```

Comment:
We don't allow either key or value as `null` inside the map because we don't allow `null` inside the MV and map is handled with MV columns

Reply:
Makes sense. I'll add `null` checks for the map keys and values so that they are not inserted into the map.